sway: 1.5 -> 1.6; wlroots: 1.12 -> 1.13

Synthetica9 opened the pull request:
###### Motivation for this change

Tracking PR for Sway 1.6, first RC just landed: https://github.com/swaywm/sway/releases/tag/1.6-rc1

EDIT 2021-03-24: Next RC landed https://github.com/swaywm/sway/releases/tag/1.6-rc2

EDIT: 2021-04-07: Final version of wlroots 0.13 landed. https://github.com/swaywm/wlroots/releases/tag/0.13.0


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Synthetica9 commented:
Seems to build and run fine here. A few new dependencies were added, I just naively added them.

primeos commented on pkgs/applications/window-managers/sway/default.nix:
```diff
@@ -33,11 +33,12 @@ stdenv.mkDerivation rec {
 
   mesonFlags = [
     "-Ddefault-wallpaper=false"
+    "-Dauto_features=auto"
```
We should avoid that. Features that we don't want should be disabled explicitly.

primeos replied:
```suggestion
    "-Dsd-bus-provider=libsystemd"
```

primeos commented on pkgs/development/libraries/wlroots/default.nix:
```diff
@@ -1,18 +1,18 @@
     repo = "wlroots";
-    rev = version;
-    sha256 = "01j38lmgs2c6fq68v8b75pkilia2wsgzgp46ivfbi9hhx47kgcfn";
+    rev = "c740fcc";
```
Nitpick (this'll be squashed later) but I'd prefer to use the full SHA-1 hashes / commit IDs here.

primeos reviewed (changes requested):
Can you also run `nixpkgs-review` and investigate broken builds (e.g. by backporting patches, switching to a newer release, looking at issue trackers, etc.)?
Other than that it looks good.

Synthetica9 committed:
labwc: unstable-2021-02-06 -> unstable-2021-03-15

Needed for wlroots 0.13.0

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

jbeich commented:
For wayfire/hikari/etc fixes try extracting from https://github.com/freebsd/freebsd-ports/commit/79579d5405fe38a2f8c8956e6f3b3f4545e058db

AndersonTorres commented on pkgs/applications/window-managers/sway/default.nix:
```diff
@@ -2,23 +2,24 @@
   patches = [
     ./sway-config-no-nix-store-references.patch
     ./load-configuration-from-etc.patch
+    ./sway-allow-wlroots-13.patch
```
Is this a Nixpkgs-only patch?

Synthetica9 replied:
Yes, but it will also be squashed away

SuperSandro2000 replied:
Such version ranges changes should use substituteInPlace.

AndersonTorres replied:
Do this via scripting, it will be easier to undo later.

AndersonTorres commented on pkgs/development/libraries/wlroots/default.nix:
```diff
@@ -23,9 +23,13 @@ stdenv.mkDerivation rec {
   ];
 
+  prePatch = ''
+    sed -i 's/0.12.0/0.13.0/' meson.build
```
Use `substituteInPlace` here.

(To be honest I also prefer sed, but using the nixpkgs-provided functions is the preferrable way.)

Synthetica9 replied:
This is going to be squashed away when wlroots 0.13 is released

AndersonTorres commented on pkgs/development/libraries/wlroots/default.nix:
```diff
@@ -23,9 +23,13 @@ stdenv.mkDerivation rec {
     libGL wayland wayland-protocols libinput libxkbcommon pixman
     xcbutilwm libX11 libcap xcbutilimage xcbutilerrors mesa
-    libpng ffmpeg
+    libpng ffmpeg libuuid xorg.xcbutilrenderutil xwayland
```
put `xcbutilrenderutil` explicitly as dependency (and use `inherit (xorg) xcbutilrenderutil;` in the callPackage).

primeos replied:
@AndersonTorres do we have a convention for that? I was a bit surprised about this but then again it shouldn't really matter anyway.

Synthetica9 commented on pkgs/applications/window-managers/labwc/default.nix:
```diff
@@ -42,6 +44,8 @@ stdenv.mkDerivation rec {
     libxkbcommon
     xwayland
+    libdrm
+    scdoc
```
Only scdoc I think?

AndersonTorres replied:
Yes, only scdoc

SuperSandro2000 commented on pkgs/applications/window-managers/dwl/default.nix:
```diff
@@ -12,8 +12,25 @@
+  totalPatches = patches ++ [
+    # Fix the renamed constant WLR_KEY_PRESSED => WL_KEYBOARD_KEY_STATE_PRESSED
+    (fetchpatch {
+      url = "https://github.com/djpohly/dwl/pull/66.patch";
```
Please don't fetch PRs but commits instead.

Synthetica9 commented:
Sway 1.6 also seems to fix an issue I was having with firefox, which is a nice plus

primeos commented on pkgs/applications/window-managers/waybox/default.nix:
```diff
@@ -41,5 +41,6 @@ stdenv.mkDerivation rec {
     license = licenses.mit;
     maintainers = with maintainers; [ AndersonTorres ];
     platforms = platforms.linux;
+    broken = true;
```
We should also notify the maintainers if we mark something as broken.
cc @AndersonTorres (already in this thread but to make it explicit)

I'd assume that `waybox` is still in use so we should either override `wlroots` (0.12) or add a TODO comment if this is only temporary and will be dropped before this PR gets merged.

Synthetica9 replied:
There doesn't seem to be any upstream development in the past half year or so, so I marked it as broken for now. I'll add a note, and I guess we'll see what the status is when the final version of wlroots 0.13 is released.

primeos replied:
Ok, that makes sense then :) Thanks!

Synthetica9 commented:
@SuperSandro2000 What's the recommended way to fetch multiple commits from a PR that hasn't been merged yet? (https://github.com/Hjdskes/cage/pull/174)

Synthetica9 commented:
@berbiche see 732cd86

primeos commented:
> @SuperSandro2000 What's the recommended way to fetch multiple commits from a PR that hasn't been merged yet? (Hjdskes/cage#174)

I think that's unfortunately a bit of a problem without a good solution (PRs are mutable so we cannot use them directly and even fetching the commits before they're merged is far from ideal as they could disappear / become GCed, e.g. if the PR branch will be rebased).

For past wlroots updates I've kindly pinged the maintainers to e.g. merge such PRs. If that doesn't work it's probably best to keep wlroots 1.12 in Nixpkgs for some time (we're already doing this with some other packages).

Synthetica9 commented:
Okay, I see you pinged the maintainers. Since this is not yet ready to be merged anyways I'll keep it as is for now, I'll use the merge commit when that appears on master.

SuperSandro2000 commented:
GCed commits is usually not a problem with github and is far less likely than a changed or.

berbiche commented:
> @berbiche see [732cd86](https://github.com/NixOS/nixpkgs/commit/732cd861f015d777756507a853d9dce806f74eda)

Cagebreak's source code history is available on the "development" branch, the master branch is just a "dump" like you said.

I'll investigate and send a patch when I have time.

jbeich commented on pkgs/applications/window-managers/labwc/default.nix:
```diff
@@ -25,11 +27,11 @@ stdenv.mkDerivation rec {
   };
 
-  nativeBuildInputs = [ pkg-config meson ninja pandoc ];
+  nativeBuildInputs = [ pkg-config meson ninja pandoc scdoc ];
```
`pandoc`  isn't used after https://github.com/johanmalm/labwc/commit/5a36fef547fc

jbeich commented on pkgs/applications/window-managers/labwc/default.nix:
```diff
@@ -25,11 +27,11 @@ stdenv.mkDerivation rec {
     repo = pname;
-    rev = "4a8fcf5c6d0b730b1e2e17e544ce7d7d3c72cd13";
-    sha256 = "g1ba8dchUN393eis0VAu1bIjQfthDGLaSijSavz4lfU=";
+    rev = "fddeb74527e5b860d9c1a91a237d390041c758b6";
```
Don't you need to bump `version` from `unstable-2021-02-06` to `unstable-2021-03-15` as well?
Alternatively, consider using `0.1.0` tag then applying https://github.com/johanmalm/labwc/pull/26.

Synthetica9 committed:
wlroots_0_12: init at 0.12.0

Uses the old wlroots in places where the new one isn't yet compatible.

Co-authored-by: Alyssa Ross <hi@alyssa.is>
Co-authored-by: Michael Weiss <dev.primeos@gmail.com>

Synthetica9 commented:
@alyssais I tried to override wlroots for wayfire and wcm, but I couldn't get it through the plugin system, any pointers?

Synthetica9 commented:
@primeos Should I do an initial squash, or wait with that until we're ready to merge?

primeos commented:
@Synthetica9 feel free to already squash/rebase it if you have time. That way we might be able to merge it a bit faster and the PR would look a bit "cleaner" but I'm fine with it either way.

Synthetica9 commented:
> feel free to already squash\rebase it if you have time.

Okay, I'll see what I can do

> **Edit: Oh, bad bot... :o :D**

🤭

primeos commented on pkgs/applications/window-managers/sway/load-configuration-from-etc.patch:
```diff
@@ -25,18 +29,20 @@ index 02b5d606..c03a9c0f 100644
+ 		{ .prefix = SYSCONFDIR, .config_folder = "sway"},
+-		{ .prefix = SYSCONFDIR, .config_folder = "i3"}
++		{ .prefix = SYSCONFDIR, .config_folder = "i3"},
++		{ .prefix = NIX_SYSCONFDIR, .config_folder = "/sway/config"},
```
```suggestion
+		{ .prefix = NIX_SYSCONFDIR, .config_folder = "sway"},
```

Didn't test this (yet) but this diff suggests it and https://github.com/swaywm/sway/commit/8f5c546fcadbacae5f9828bc1bfac247186cfdfc confirms it (the new `config_path` function).

alyssais commented:
> @alyssais I tried to override wlroots for wayfire and wcm, but I
> couldn't get it through the plugin system, any pointers?

I've just pushed a commit to this PR that should do this -- do with it
whatever you want. :)

BTW: in my experience usual convention is to call versioned attributes
e.g. wlroots_0_12, rather than wlroots-0-12.

alyssais commented:
I pushed a small cleanup to the Wayfire override.
This is all going to be squashed a bit before being merged, right?

Synthetica9 commented:
> This is all going to be squashed a bit before being merged, right?

Yeah, I'll do a rebase to group relevant commits together at least, maybe a full squash if requested.

mweinelt commented:
https://github.com/swaywm/sway/releases/tag/1.6

Synthetica9 commented:
> [swaywm/sway@`1.6` (release)](https://github.com/swaywm/sway/releases/tag/1.6)

Way ahead of you 😉

Synthetica9 commented:
@primeos I think two things remain: 

- [ ] What do we do with https://github.com/Hjdskes/cage/pull/174 ? Pin wlroots to 0.12 for cage? 
- [ ] Rebase again, should I squash all RCs together or keep them for the "story"?

Synthetica9 committed:
wlroots{,_0_12}: add synthetica (myself) as maintainer

primeos commented:
@Synthetica9 yes, I'd pin cage to wlroots 0.12 (we can always look at that later, outside of this PR) and squash the RCs together. Regarding squashing I'd do the following (IIRC that's how I handled it in the past but that is open for discussion if you don't like it):
1. Commit: Introduces `wlroots_0_12` and pins the packages that still require it to that version
2. Commit: Updates `wlroots` and introduces the required changes to fix builds of other packages (except if we update their version)
3. Commit: Updates `sway`

Synthetica9 committed:
wlroots: 0.12.0 -> 0.13.0

Pulls in an upstream patch for dwl. Explicitly takes xcbutilrenderutil
as an argument to avoid pulling in all of xorg.

Synthetica9 committed:
sway: 1.5.1 -> 1.6

Co-authored-by: Michael Weiss <dev.primeos@gmail.com>

Synthetica9 commented:
I think this PR is ready to be merged :)

primeos commented on pkgs/applications/window-managers/cage/default.nix:
```diff
@@ -4,6 +4,7 @@
 , systemd, libGL, libX11
 , xwayland ? null
 , nixosTests
+, fetchpatch
```
Nitpick but could you drop this leftover if you don't mind?

primeos commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26432,10 +26440,12 @@ in
   inherit (wayfireApplications) wayfire wcm;
-  wayfireApplications-unwrapped = recurseIntoAttrs (callPackage ../applications/window-managers/wayfire/applications.nix { });
-  wayfirePlugins = recurseIntoAttrs (callPackage ../applications/window-managers/wayfire/plugins.nix {
+  wayfireApplications-unwrapped =
+    (callPackage ../applications/window-managers/wayfire/applications.nix { }).
+    extend (_: _: { wlroots = wlroots_0_12; });
+  wayfirePlugins = callPackage ../applications/window-managers/wayfire/plugins.nix {
```
We should re-add `recurseIntoAttrs` for `wayfireApplications-unwrapped` and `wayfirePlugins` (relevant for nix-env, the package search, etc.). That should also fix this nixpkgs-review output:
```
3 packages removed:
wayfire (†0.7.0) wcm (†0.7.0) wf-shell (†0.7.0)
```

primeos replied:
If you don't mind you could also drop `being cage and wayfire.` from this commit message (there are more affected packages now but we don't need to name them in the commit message).

(The Wayfire change could be an extra commit as we're also making it extensible but that's likely not worth the effort so feel free to ignore that, depending on what you prefer.)

Synthetica9 replied:
@primeos I have pushed something that compiles, checking the result of nixpkgs-review now

Synthetica9 replied:
Hmm, now it still says 

```
1 package removed:
wf-shell (†0.7.0)
```

Not sure why...

primeos replied:
It still missing for `wayfirePlugins` (will appear as `wayfireApplications-unwrapped.wayfirePlugins.wf-shell` in the search). Then it should work.

Synthetica9 replied:
Oh, I see... Should it be on both or only on unwrapped?

primeos replied:
It used to be on both `wayfireApplications-unwrapped` and `wayfirePlugins` (s. diff) which is why I'd keep it on both.

Synthetica9 replied:
Okay. Marking this resolved 👍🏻

primeos reviewed (changes requested):
Overall it LGTM now, thanks a lot! We mainly need to re-add `recurseIntoAttrs` for Wayfire (s. below).

`nixpkgs-review` result:
```
1 package added:
wlroots_0_12 (init at 0.12.0)

11 packages updated:
dwl fuzzel sway-contrib.grimshot (1.5.1 → 1.6) labwc (2021-02-06 → 2021-03-15) qt-video-wlr sway (1.5.1 → 1.6) sway-contrib.inactive-windows-transparency (1.5.1 → 1.6) sway-unwrapped (1.5.1 → 1.6) waybar wbg-unstable wlroots (0.12.0 → 0.13.0)

3 packages removed:
wayfire (†0.7.0) wcm (†0.7.0) wf-shell (†0.7.0)

12 packages built:
dwl fuzzel labwc qt-video-wlr sway sway-contrib.grimshot sway-contrib.inactive-windows-transparency sway-unwrapped waybar wbg wlroots wlroots_0_12
```

primeos reviewed (approved):
Thanks a lot, very well done!
I've briefly tested `sway`, `dwl`, and `labwc` on my old desktop PC. `nixpkgs-review` result:
```
1 package added:
wayfireApplications-unwrapped.wlroots (init at 0.12.0)

11 packages updated:
dwl fuzzel sway-contrib.grimshot (1.5.1 → 1.6) labwc (2021-02-06 → 2021-03-15) qt-video-wlr sway (1.5.1 → 1.6) sway-contrib.inactive-windows-transparency (1.5.1 → 1.6) sway-unwrapped (1.5.1 → 1.6) waybar wbg-unstable wlroots (0.12.0 → 0.13.0)

12 packages built:
dwl fuzzel labwc qt-video-wlr sway sway-contrib.grimshot sway-contrib.inactive-windows-transparency sway-unwrapped waybar wayfireApplications-unwrapped.wlroots wbg wlroots
```
